dao/seeding.go

```diff
@@ -0,0 +1,185 @@
+package dao
+
+import (
+	"encoding/json"
+	"os"
+
+	l "github.com/RedHatInsights/sources-api-go/logger"
+	m "github.com/RedHatInsights/sources-api-go/model"
+	"sigs.k8s.io/yaml"
+)
+
+const SEEDS_DIR = "./dao/seeds/"
+
+type (
+	sourceTypeSeed      map[string]m.SourceTypeSeed
+	applicationTypeSeed map[string]m.ApplicationTypeSeed
+	metadataSeed        map[string]m.MetaDataSeed
+)
+
+func SeedDatabase() error {
+	err := seedSourceTypes()
+	if err != nil {
+		return err
+	}
+	err = seedApplicationTypes()
+	if err != nil {
+		return err
+	}
+	err = seedMetaData()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func seedSourceTypes() error {
+	// parse the map of seeds
+	seeds := make(sourceTypeSeed)
+	data, err := os.ReadFile(SEEDS_DIR + "source_types.yml")
+	if err != nil {
+		return err
+	}
+
+	err = yaml.Unmarshal(data, &seeds)
+	if err != nil {
+		return err
+	}
+
+	// load all the seeds from the db
+	sourceTypes := make([]m.SourceType, 0)
+	result := DB.Model(&m.SourceType{}).Scan(&sourceTypes)
+	if result.Error != nil {
+		return result.Error
+	}
+
+	// loop through all of the seeds
+	// 1. creating the record if they don't exist
+	// 2. updating the ones that do exist
+	// 3. deleting the ones removed from the list
+	for name, values := range seeds {
+		var st *m.SourceType
+
+		// find an existing one in the list of source types
+		for i, sourceType := range sourceTypes {
```

Comment:
@lindgrenj6 

Do we need cycle inside ? 
What about to fetch all sources to map(let's say in variable sourceTypeMap) `map[string]m.SourceType` where index is name of source type - before this `seeds` loop.

I guess then we can ask check  `if sourceTypeMap[name] != nil` and (as you have):
1. If exists then -  update and remove from(maybe set nil) `sourceTypeMap[name]`
2. if don't exists - create 
3. What is remaining in sourceTypeMap - it should be removed.

Let me what do you think but It looks with this we can avoid this loop inside other loop.

Reply:
Ok I went through and changed this - there are 2 new functions that parse the list of existing SourceTypes/ApplicationTypes into a `map[string]*SourceType`.

This way we can check if something exists or not (map will have a nil value for non-existing attributes), and we just delete them out of the hash instead of deleting them out of the slice.

It is a bit longer code-wise but it is pretty straight forward still.